File: 14次/t4 Runge-Kutta method/test4.py
import matplotlib.pyplot as plt
import numpy as np
def pqt(pq):
    p=pq[0];
    q=pq[1]
    temp=(p**2+q**2);
    return np.array([-temp*q,temp*p])
#龙格库塔

def runge_kutta(pi,qi,h):
    #对于x和y的认识
    #array与[]运算限制
    pq=np.array([pi,qi])
    k1=pqt(pq);
    k2=pqt(pq+h/2*k1);
    k3=pqt(pq+h/2*k2);
    k4=pqt(pq+h*k3);
    
    pqi=pq+h/6*(k1+2*k2+2*k3+k4)
    return pqi[0],pqi[1]


#龙格库塔
p=1;q=0

N=500000;
h=0.1;
x0=0;
x=[x0+h*i for i in range(0,N+1)];
p=np.zeros(N+1);
q=np.zeros(N+1)
p[0]=1;q[0]=0

for i in range(0,N):
    p[i+1],q[i+1]=runge_kutta(p[i],q[i],h);
    # print(p[i+1])

plt.plot(p,q)

plt.xlabel('p',fontsize=20);
plt.ylabel('q',fontsize=20);
plt.title('Runge-Kutta method',fontsize=20);
plt.tick_params(labelsize=20)
R=p**2+q**2;
maxR=max(R)
minR=min(R)
print("最大半径",maxR)
print("最小半径",minR)
print("半径误差",maxR-minR)

File: 14次/t4 Runge-Kutta method/test4.m
function main()


T=500000;
tau=0.1;

p=zeros(T,1);
q=zeros(T,1);

p(1)=1;
q(1)=0;

N=1;
yn=[1;0];

for t=2:T*N
    
    y0=yn;
    
    k1=f(y0);
    k2=f(y0+1/2*tau*k1);
    k3=f(y0+1/2*tau*k2);
    k4=f(y0+tau*k3);
    
    
    
    yn=y0+tau/6*(k1+2*k2+2*k3+k4);
    
    if(mod(t,N)==0)
        p(t/N)=yn(1);
        q(t/N)=yn(2);
    end
    
end


plot(p,q)

xlabel('p','Fontsize', 20);
ylabel('q','Fontsize', 20);
title('Runge-Kutta method','Fontsize', 20);
set(gca, 'Fontsize', 20);

R=p.^2+q.^2;
maxR=max(R)
minR=min(R)

errorR=maxR-minR


end

function y=f(y0)

    p=y0(1);
    q=y0(2);
    y=zeros(2,1);
    y(1)=-(p^2+q^2)*q;
    y(2)=(p^2+q^2)*p;

end
